File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(1) do
    """
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    """
  end

  def verse(2) do
    number = 1
    """
    #{bottle(number)} of beer on the wall, #{bottle(number)} of beer.
    Take it down and pass it around, no more bottles of beer on the wall.
    """
  end

  def verse(number) do
    number = number - 1
    """
    #{bottle(number)} of beer on the wall, #{bottle(number)} of beer.
    Take one down and pass it around, #{bottle(number - 1)} of beer on the wall.
    """
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range \\ 100..1) do
    Enum.map_join(range, "\n", &verse/1)
  end

  def bottle(1), do: "1 bottle"
  def bottle(number), do: "#{number} bottles"
end

File: ecmascript/beer-song/beer-song.js
const phrase0 = () => { return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n" };
const phrase1 = () => { return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n" };
const phraseN = (n) => { return `${n} bottles of beer on the wall, ${n} bottles of beer.\nTake one down and pass it around, ${n - 1} bottle${n - 1 == 1 ? '' : 's'} of beer on the wall.\n` };

class Beer {
  static verse(number) {
    switch (number) {
      case 0: return phrase0();
      case 1: return phrase1();
      default: return phraseN(number);
    }
  }

  static sing(start = 99, end = 0) {
    let out = [];
    for (var x = start; x >= end; x--) {
      out.push(this.verse(x));
    }
    return out.join("\n");
  };
}

export default Beer;